README.md
# What is RudderStack?

[RudderStack](https://rudderstack.com/) is a **customer data pipeline** tool for collecting, routing and processing data from your websites, apps, cloud tools, and data warehouse.

More information on RudderStack can be found [here](https://github.com/rudderlabs/rudder-server).

## Integrating Branch with RudderStack's Android SDK

1. Add [BranchIO](https://branch.io) as a destination in the [Dashboard](https://app.rudderlabs.com/) and define ```branchKey```

2. Add the dependency under ```dependencies```
```
implementation 'com.rudderstack.android.sdk:core:[1.20.1, 2.0.0)'
implementation 'com.rudderstack.android.integration:branch:1.0.0'
```

3. Add the following optional dependencies needed by Branch under ```dependencies```

```
// required if your app is in the Google Play Store (tip: avoid using bundled play services libs)
implementation 'com.google.android.gms:play-services-ads-identifier:17.1.0+'
// alternatively, use the following lib for getting the AAID
// implementation 'com.google.android.gms:play-services-ads:17.2.0'
// optional
// Chrome Tab matching (enables 100% guaranteed matching based on cookies)
implementation 'androidx.browser:browser:1.0.0'
// Replace above with the line below if you do not support androidx
// implementation 'com.android.support:customtabs:28.0.0'
```

## Initialize ```RudderClient```
```
val rudderClient: RudderClient = RudderClient.getInstance(
    this,
    <WRITE_KEY>,
    RudderConfig.Builder()
        .withDataPlaneUrl(<DATA_PLANE_URL>)
        .withLogLevel(RudderLogger.RudderLogLevel.DEBUG)
        .withFactory(BranchIntegrationFactory.FACTORY)
        .build()
)
```

## Send Events

Follow the steps from the [RudderStack Android SDK](https://github.com/rudderlabs/rudder-sdk-android).

## Contact Us

If you come across any issues while configuring or using this integration, please feel free to start a conversation on our [Slack](https://resources.rudderstack.com/join-rudderstack-slack) channel. We will be happy to help you.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## 2.0.0 (2023-12-05)


### ⚠ BREAKING CHANGES

* upgrade branch sdk to v5.7.2 (#19)

### Features

* upgrade branch sdk to v5.7.2 ([#19](https://github.com/rudderlabs/rudder-integration-branch-android/issues/19)) ([8ca9bc8](https://github.com/rudderlabs/rudder-integration-branch-android/commit/8ca9bc802c2700b00b655e8c5dac4d0f2cee33ec)), closes [#14](https://github.com/rudderlabs/rudder-integration-branch-android/issues/14)


branch/src/test/java/com/rudderstack/android/integration/branch/TestUtils.kt
package com.rudderstack.android.integration.branch

import com.google.gson.GsonBuilder
import com.rudderstack.android.sdk.core.RudderMessage
import io.branch.referral.util.ContentMetadata
import io.branch.referral.util.CurrencyType
import io.branch.referral.util.ProductCategory
import io.mockk.CapturingSlot
import org.junit.Assert
import java.io.BufferedReader
import java.io.IOException
import java.io.InputStream
import java.io.InputStreamReader
import java.util.ArrayList

class TestUtils {
    private val gson = GsonBuilder().create()

    fun getMessage(fileName: String): RudderMessage {
        val inputJson = getJsonFromPath(fileName)
        return parseJson(inputJson!!)
    }

    fun getJsonFromPath(path: String?): String? {
        val inputStream: InputStream =
            this.javaClass.classLoader?.getResourceAsStream(path) ?: return null
        val reader = BufferedReader(InputStreamReader(inputStream))
        val builder = StringBuilder()
        var line: String?
        try {
            while (reader.readLine().also { line = it } != null) {
                builder.append(line)
            }
        } catch (ex: IOException) {
            ex.printStackTrace()
            return null
        }
        return builder.toString()
    }

    private fun parseJson(json: String): RudderMessage {
        return gson.fromJson(json, RudderMessage::class.java)
    }

    fun verifySingleProductMetadata(contentMetadata: ContentMetadata) {
        Assert.assertEquals(154.79, contentMetadata.price, 0.0)
        Assert.assertEquals(CurrencyType.USD, contentMetadata.currencyType)
        Assert.assertEquals("some_brand", contentMetadata.productBrand)
        Assert.assertEquals("some_name", contentMetadata.productName)
        Assert.assertEquals(
            ProductCategory.ANIMALS_AND_PET_SUPPLIES,
            contentMetadata.productCategory
        )
        Assert.assertEquals("some_sku", contentMetadata.sku)
        Assert.assertEquals(10.0, contentMetadata.quantity, 0.0)
        Assert.assertEquals("some_variant", contentMetadata.productVariant)
        Assert.assertEquals(4.5, contentMetadata.rating, 0.0)
    }

    fun verifyStandardProperties(
        affiliationSlot: CapturingSlot<String>,
        currencySlot: CapturingSlot<CurrencyType>,
        couponSlot: CapturingSlot<String>,
        revenueSlot: CapturingSlot<Double>,
        shippingSlot: CapturingSlot<Double>,
        taxSlot: CapturingSlot<Double>,
        orderIdSlot: CapturingSlot<String>
    ) {
        Assert.assertEquals("some_affiliation", affiliationSlot.captured)
        Assert.assertEquals(CurrencyType.USD, currencySlot.captured)
        Assert.assertEquals("some_coupon", couponSlot.captured)
        Assert.assertEquals(754.79, revenueSlot.captured, 0.0)
        Assert.assertEquals(100.0, shippingSlot.captured, 0.0)
        Assert.assertEquals(5.0, taxSlot.captured, 0.0)
        Assert.assertEquals("some_order_id", orderIdSlot.captured)
    }

    fun verifyProduct1(product1: ContentMetadata) {
        Assert.assertEquals(154.79, product1.price, 0.0)
        Assert.assertEquals(CurrencyType.USD, product1.currencyType)
        Assert.assertEquals("some_brand_1", product1.productBrand)
        Assert.assertEquals("some_name_1", product1.productName)
        Assert.assertEquals(ProductCategory.ANIMALS_AND_PET_SUPPLIES, product1.productCategory)
        Assert.assertEquals("some_sku_1", product1.sku)
        Assert.assertEquals(10.0, product1.quantity, 0.0)
        Assert.assertEquals("some_variant_1", product1.productVariant)
        Assert.assertEquals(4.5, product1.rating, 0.0)
    }

    fun verifyProduct2(product1: ContentMetadata) {
        Assert.assertEquals(254.79, product1.price, 0.0)
        Assert.assertEquals(CurrencyType.INR, product1.currencyType)
        Assert.assertEquals("some_brand_2", product1.productBrand)
        Assert.assertEquals("some_name_2", product1.productName)
        Assert.assertEquals(ProductCategory.ANIMALS_AND_PET_SUPPLIES, product1.productCategory)
        Assert.assertEquals("some_sku_2", product1.sku)
        Assert.assertEquals(20.0, product1.quantity, 0.0)
        Assert.assertEquals("some_variant_2", product1.productVariant)
        Assert.assertEquals(3.5, product1.rating, 0.0)
    }

    fun verifyCustomProperty(
        propertyNameSlot: MutableList<String>,
        propertyValueSlot: MutableList<String>
    ) {
        Assert.assertEquals(mutableListOf("key-1", "key-2"), propertyNameSlot)
        Assert.assertEquals(mutableListOf("value-1", "123.0"), propertyValueSlot)
    }

    fun verifyQueryProperty(captureBranchUniversalObject: ArrayList<String>) {
        Assert.assertEquals("some_query", captureBranchUniversalObject[0])
    }

    fun getUserIdOfLength(length: Int): String? {
        val sb = StringBuilder()
        for (j in 0 until length - 1) {
            sb.append("a")
        }
        sb.append(".")
        return sb.toString()
    }

    fun verifyIdentify(userId: String) {
        Assert.assertEquals("some_user_id", userId)
    }

    fun verifyIdentifyWhenUserIdIsTruncated(userId: String) {
        Assert.assertEquals("aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
            userId)
    }
}

branch/src/test/java/com/rudderstack/android/integration/branch/BranchIntegrationFactoryUnitTest.kt
package com.rudderstack.android.integration.branch

import android.app.Application
import io.branch.indexing.BranchUniversalObject
import io.branch.referral.Branch
import io.branch.referral.util.BRANCH_STANDARD_EVENT
import io.branch.referral.util.BranchEvent
import io.branch.referral.util.CurrencyType
import io.mockk.OfTypeMatcher
import io.mockk.every
import io.mockk.mockk
import io.mockk.mockkConstructor
import io.mockk.slot
import org.junit.Before
import org.junit.Test

class BranchIntegrationFactoryUnitTest {
    val testUtils = TestUtils()
    private var branchIntegrationFactory: BranchIntegrationFactory? = null

    private val branchInstance = mockk<Branch>()
    private val applicationContext = mockk<Application>()

    @Before
    fun setUp() {
        this.branchIntegrationFactory = BranchIntegrationFactory(branchInstance, applicationContext)
    }

    @Test
    fun testProductAdded() {
        val branchUniversalObjectSlot = slot<BranchUniversalObject>()
        val propertyNameSlot = mutableListOf<String>()
        val propertyValueSlot = mutableListOf<String>()
        val mockBranchEvent = mockk<BranchEvent>()

        mockkConstructor(BranchEvent::class)
        every { constructedWith<BranchEvent>(OfTypeMatcher<BRANCH_STANDARD_EVENT>(BRANCH_STANDARD_EVENT::class)).addContentItems(capture(branchUniversalObjectSlot)) } returns mockBranchEvent
        every { constructedWith<BranchEvent>(OfTypeMatcher<BRANCH_STANDARD_EVENT>(BRANCH_STANDARD_EVENT::class)).addCustomDataProperty(capture(propertyNameSlot), capture(propertyValueSlot)) } returns mockBranchEvent

        val message = testUtils.getMessage("productAdded.json")
        branchIntegrationFactory?.dump(message)

        val captureBranchUniversalObject = branchUniversalObjectSlot.captured
        testUtils.verifySingleProductMetadata(captureBranchUniversalObject.contentMetadata)
        testUtils.verifyCustomProperty(propertyNameSlot, propertyValueSlot)
    }

    @Test
    fun testProductAddedToWishList() {
        val propertyNameSlot = mutableListOf<String>()
        val propertyValueSlot = mutableListOf<String>()
        val mockBranchEvent = mockk<BranchEvent>()

        mockkConstructor(BranchEvent::class)
        every { constructedWith<BranchEvent>(OfTypeMatcher<BRANCH_STANDARD_EVENT>(BRANCH_STANDARD_EVENT::class)).addCustomDataProperty(capture(propertyNameSlot), capture(propertyValueSlot)) } returns mockBranchEvent

        val message = testUtils.getMessage("productAddedToWishList.json")
        branchIntegrationFactory?.dump(message)

        testUtils.verifyCustomProperty(propertyNameSlot, propertyValueSlot)
    }

    @Test
    fun testCartViewed() {
        val listOfBranchUniversalObjectSlot = slot<List<BranchUniversalObject>>()
        val propertyNameSlot = mutableListOf<String>()
        val propertyValueSlot = mutableListOf<String>()
        val mockBranchEvent = mockk<BranchEvent>()

        val affiliationSlot = slot<String>()
        val currencySlot = slot<CurrencyType>()
        val couponSlot = slot<String>()
        val revenueSlot = slot<Double>()
        val shippingSlot = slot<Double>()
        val taxSlot = slot<Double>()
        val orderIdSlot = slot<String>()

        mockkConstructor(BranchEvent::class)
        every { constructedWith<BranchEvent>(OfTypeMatcher<BRANCH_STANDARD_EVENT>(BRANCH_STANDARD_EVENT::class)).addContentItems(capture(listOfBranchUniversalObjectSlot)) } returns mockBranchEvent
        every { constructedWith<BranchEvent>(OfTypeMatcher<BRANCH_STANDARD_EVENT>(BRANCH_STANDARD_EVENT::class)).addCustomDataProperty(capture(propertyNameSlot), capture(propertyValueSlot)) } returns mockBranchEvent

        every { constructedWith<BranchEvent>(OfTypeMatcher<BRANCH_STANDARD_EVENT>(BRANCH_STANDARD_EVENT::class)).setAffiliation(capture(affiliationSlot)) } returns mockBranchEvent
        every { constructedWith<BranchEvent>(OfTypeMatcher<BRANCH_STANDARD_EVENT>(BRANCH_STANDARD_EVENT::class)).setCurrency(capture(currencySlot)) } returns mockBranchEvent
        every { constructedWith<BranchEvent>(OfTypeMatcher<BRANCH_STANDARD_EVENT>(BRANCH_STANDARD_EVENT::class)).setCoupon(capture(couponSlot)) } returns mockBranchEvent
        every { constructedWith<BranchEvent>(OfTypeMatcher<BRANCH_STANDARD_EVENT>(BRANCH_STANDARD_EVENT::class)).setRevenue(capture(revenueSlot)) } returns mockBranchEvent
        every { constructedWith<BranchEvent>(OfTypeMatcher<BRANCH_STANDARD_EVENT>(BRANCH_STANDARD_EVENT::class)).setShipping(capture(shippingSlot)) } returns mockBranchEvent
        every { constructedWith<BranchEvent>(OfTypeMatcher<BRANCH_STANDARD_EVENT>(BRANCH_STANDARD_EVENT::class)).setTax(capture(taxSlot)) } returns mockBranchEvent
        every { constructedWith<BranchEvent>(OfTypeMatcher<BRANCH_STANDARD_EVENT>(BRANCH_STANDARD_EVENT::class)).setTransactionID(capture(orderIdSlot)) } returns mockBranchEvent

        val message = testUtils.getMessage("cartViewed.json")
        branchIntegrationFactory?.dump(message)

        val captureBranchUniversalObject = listOfBranchUniversalObjectSlot.captured
        testUtils.verifyStandardProperties(affiliationSlot, currencySlot, couponSlot, revenueSlot, shippingSlot, taxSlot, orderIdSlot)
        testUtils.verifyProduct1(captureBranchUniversalObject[0].getContentMetadata())
        testUtils.verifyProduct2(captureBranchUniversalObject[1].getContentMetadata())
        testUtils.verifyCustomProperty(propertyNameSlot, propertyValueSlot)
    }

    @Test
    fun testProductsSearched() {
        val branchUniversalObjectSlot = slot<BranchUniversalObject>()
        val propertyNameSlot = mutableListOf<String>()
        val propertyValueSlot = mutableListOf<String>()
        val mockBranchEvent = mockk<BranchEvent>()

        mockkConstructor(BranchEvent::class)
        every { constructedWith<BranchEvent>(OfTypeMatcher<BRANCH_STANDARD_EVENT>(BRANCH_STANDARD_EVENT::class)).addContentItems(capture(branchUniversalObjectSlot)) } returns mockBranchEvent
        every { constructedWith<BranchEvent>(OfTypeMatcher<BRANCH_STANDARD_EVENT>(BRANCH_STANDARD_EVENT::class)).addCustomDataProperty(capture(propertyNameSlot), capture(propertyValueSlot)) } returns mockBranchEvent

        val message = testUtils.getMessage("productsSearched.json")
        branchIntegrationFactory?.dump(message)

        val captureBranchUniversalObject = branchUniversalObjectSlot.captured
        testUtils.verifyQueryProperty(captureBranchUniversalObject.keywords)
        testUtils.verifyCustomProperty(propertyNameSlot, propertyValueSlot)
    }

    @Test
    fun testCustomTrackEvent() {
        val propertyNameSlot = mutableListOf<String>()
        val propertyValueSlot = mutableListOf<String>()
        val mockBranchEvent = mockk<BranchEvent>()

        mockkConstructor(BranchEvent::class)
        every { constructedWith<BranchEvent>(OfTypeMatcher<String>(String::class)).addCustomDataProperty(capture(propertyNameSlot), capture(propertyValueSlot)) } returns mockBranchEvent

        val message = testUtils.getMessage("customTrackEvent.json")
        branchIntegrationFactory?.dump(message)

        testUtils.verifyCustomProperty(propertyNameSlot, propertyValueSlot)
    }

    @Test
    fun testIdentifyEvent() {
        val userIdSlot = slot<String>()

        every { branchInstance.setIdentity(capture(userIdSlot)) } returns Unit

        val message = testUtils.getMessage("identify.json")
        branchIntegrationFactory?.dump(message)

        testUtils.verifyIdentify(userIdSlot.captured)
    }

    @Test
    fun testIdentifyEventWith128CharacterUserId() {
        val userIdSlot = slot<String>()

        every { branchInstance.setIdentity(capture(userIdSlot)) } returns Unit

        val message = testUtils.getMessage("identifyWith128CharacterUserId.json")
        branchIntegrationFactory?.dump(message)

        testUtils.verifyIdentifyWhenUserIdIsTruncated(userIdSlot.captured)
    }
}

branch/src/test/java/com/rudderstack/android/integration/branch/ComScoreIntegrationFactoryUnitTest.kt
package com.rudderstack.android.integration.branch

import android.app.Application
import io.branch.indexing.BranchUniversalObject
import io.branch.referral.Branch
import io.branch.referral.util.BRANCH_STANDARD_EVENT
import io.branch.referral.util.BranchEvent
import io.branch.referral.util.CurrencyType
import io.mockk.OfTypeMatcher
import io.mockk.every
import io.mockk.mockk
import io.mockk.mockkConstructor
import io.mockk.slot
import org.junit.Before
import org.junit.Test

class ComScoreIntegrationFactoryUnitTest {
    val testUtils = TestUtils()
    private var branchIntegrationFactory: BranchIntegrationFactory? = null

    private val branchInstance = mockk<Branch>()
    private val applicationContext = mockk<Application>()

    @Before
    fun setUp() {
        this.branchIntegrationFactory = BranchIntegrationFactory(branchInstance, applicationContext)
    }

    @Test
    fun testProductAdded() {
        val branchUniversalObjectSlot = slot<BranchUniversalObject>()
        val propertyNameSlot = mutableListOf<String>()
        val propertyValueSlot = mutableListOf<String>()
        val mockBranchEvent = mockk<BranchEvent>()

        mockkConstructor(BranchEvent::class)
        every { constructedWith<BranchEvent>(OfTypeMatcher<BRANCH_STANDARD_EVENT>(BRANCH_STANDARD_EVENT::class)).addContentItems(capture(branchUniversalObjectSlot)) } returns mockBranchEvent
        every { constructedWith<BranchEvent>(OfTypeMatcher<BRANCH_STANDARD_EVENT>(BRANCH_STANDARD_EVENT::class)).addCustomDataProperty(capture(propertyNameSlot), capture(propertyValueSlot)) } returns mockBranchEvent

        val message = testUtils.getMessage("productAdded.json")
        branchIntegrationFactory?.dump(message)

        val captureBranchUniversalObject = branchUniversalObjectSlot.captured
        testUtils.verifySingleProductMetadata(captureBranchUniversalObject.contentMetadata)
        testUtils.verifyCustomProperty(propertyNameSlot, propertyValueSlot)
    }

    @Test
    fun testProductAddedToWishList() {
        val propertyNameSlot = mutableListOf<String>()
        val propertyValueSlot = mutableListOf<String>()
        val mockBranchEvent = mockk<BranchEvent>()

        mockkConstructor(BranchEvent::class)
        every { constructedWith<BranchEvent>(OfTypeMatcher<BRANCH_STANDARD_EVENT>(BRANCH_STANDARD_EVENT::class)).addCustomDataProperty(capture(propertyNameSlot), capture(propertyValueSlot)) } returns mockBranchEvent

        val message = testUtils.getMessage("productAddedToWishList.json")
        branchIntegrationFactory?.dump(message)

        testUtils.verifyCustomProperty(propertyNameSlot, propertyValueSlot)
    }

    @Test
    fun testCartViewed() {
        val listOfBranchUniversalObjectSlot = slot<List<BranchUniversalObject>>()
        val propertyNameSlot = mutableListOf<String>()
        val propertyValueSlot = mutableListOf<String>()
        val mockBranchEvent = mockk<BranchEvent>()

        val affiliationSlot = slot<String>()
        val currencySlot = slot<CurrencyType>()
        val couponSlot = slot<String>()
        val revenueSlot = slot<Double>()
        val shippingSlot = slot<Double>()
        val taxSlot = slot<Double>()
        val orderIdSlot = slot<String>()

        mockkConstructor(BranchEvent::class)
        every { constructedWith<BranchEvent>(OfTypeMatcher<BRANCH_STANDARD_EVENT>(BRANCH_STANDARD_EVENT::class)).addContentItems(capture(listOfBranchUniversalObjectSlot)) } returns mockBranchEvent
        every { constructedWith<BranchEvent>(OfTypeMatcher<BRANCH_STANDARD_EVENT>(BRANCH_STANDARD_EVENT::class)).addCustomDataProperty(capture(propertyNameSlot), capture(propertyValueSlot)) } returns mockBranchEvent

        every { constructedWith<BranchEvent>(OfTypeMatcher<BRANCH_STANDARD_EVENT>(BRANCH_STANDARD_EVENT::class)).setAffiliation(capture(affiliationSlot)) } returns mockBranchEvent
        every { constructedWith<BranchEvent>(OfTypeMatcher<BRANCH_STANDARD_EVENT>(BRANCH_STANDARD_EVENT::class)).setCurrency(capture(currencySlot)) } returns mockBranchEvent
        every { constructedWith<BranchEvent>(OfTypeMatcher<BRANCH_STANDARD_EVENT>(BRANCH_STANDARD_EVENT::class)).setCoupon(capture(couponSlot)) } returns mockBranchEvent
        every { constructedWith<BranchEvent>(OfTypeMatcher<BRANCH_STANDARD_EVENT>(BRANCH_STANDARD_EVENT::class)).setRevenue(capture(revenueSlot)) } returns mockBranchEvent
        every { constructedWith<BranchEvent>(OfTypeMatcher<BRANCH_STANDARD_EVENT>(BRANCH_STANDARD_EVENT::class)).setShipping(capture(shippingSlot)) } returns mockBranchEvent
        every { constructedWith<BranchEvent>(OfTypeMatcher<BRANCH_STANDARD_EVENT>(BRANCH_STANDARD_EVENT::class)).setTax(capture(taxSlot)) } returns mockBranchEvent
        every { constructedWith<BranchEvent>(OfTypeMatcher<BRANCH_STANDARD_EVENT>(BRANCH_STANDARD_EVENT::class)).setTransactionID(capture(orderIdSlot)) } returns mockBranchEvent

        val message = testUtils.getMessage("cartViewed.json")
        branchIntegrationFactory?.dump(message)

        val captureBranchUniversalObject = listOfBranchUniversalObjectSlot.captured
        testUtils.verifyStandardProperties(affiliationSlot, currencySlot, couponSlot, revenueSlot, shippingSlot, taxSlot, orderIdSlot)
        testUtils.verifyProduct1(captureBranchUniversalObject[0].getContentMetadata())
        testUtils.verifyProduct2(captureBranchUniversalObject[1].getContentMetadata())
        testUtils.verifyCustomProperty(propertyNameSlot, propertyValueSlot)
    }

    @Test
    fun testProductsSearched() {
        val branchUniversalObjectSlot = slot<BranchUniversalObject>()
        val propertyNameSlot = mutableListOf<String>()
        val propertyValueSlot = mutableListOf<String>()
        val mockBranchEvent = mockk<BranchEvent>()

        mockkConstructor(BranchEvent::class)
        every { constructedWith<BranchEvent>(OfTypeMatcher<BRANCH_STANDARD_EVENT>(BRANCH_STANDARD_EVENT::class)).addContentItems(capture(branchUniversalObjectSlot)) } returns mockBranchEvent
        every { constructedWith<BranchEvent>(OfTypeMatcher<BRANCH_STANDARD_EVENT>(BRANCH_STANDARD_EVENT::class)).addCustomDataProperty(capture(propertyNameSlot), capture(propertyValueSlot)) } returns mockBranchEvent

        val message = testUtils.getMessage("productsSearched.json")
        branchIntegrationFactory?.dump(message)

        val captureBranchUniversalObject = branchUniversalObjectSlot.captured
        testUtils.verifyQueryProperty(captureBranchUniversalObject.keywords)
        testUtils.verifyCustomProperty(propertyNameSlot, propertyValueSlot)
    }

    @Test
    fun testCustomTrackEvent() {
        val propertyNameSlot = mutableListOf<String>()
        val propertyValueSlot = mutableListOf<String>()
        val mockBranchEvent = mockk<BranchEvent>()

        mockkConstructor(BranchEvent::class)
        every { constructedWith<BranchEvent>(OfTypeMatcher<String>(String::class)).addCustomDataProperty(capture(propertyNameSlot), capture(propertyValueSlot)) } returns mockBranchEvent

        val message = testUtils.getMessage("customTrackEVent.json")
        branchIntegrationFactory?.dump(message)

        testUtils.verifyCustomProperty(propertyNameSlot, propertyValueSlot)
    }

    @Test
    fun testIdentifyEvent() {
        val userIdSlot = slot<String>()
        val mockBranchEvent = mockk<BranchEvent>()

        every { branchInstance.setIdentity(capture(userIdSlot)) } returns Unit

        val message = testUtils.getMessage("identify.json")
        branchIntegrationFactory?.dump(message)

        testUtils.verifyIdentify(userIdSlot.captured)
    }

    @Test
    fun testIdentifyEventWith128CharacterUserId() {
        val userIdSlot = slot<String>()
        val mockBranchEvent = mockk<BranchEvent>()

        every { branchInstance.setIdentity(capture(userIdSlot)) } returns Unit

        val message = testUtils.getMessage("identifyWith128CharacterUserId.json")
        val userId = testUtils.getUserIdOfLength(128)
        branchIntegrationFactory?.dump(message)

        testUtils.verifyIdentifyWhenUserIdIsTruncated(userIdSlot.captured)
    }
}

sample-kotlin/src/main/java/com/rudderlabs/android/sample/kotlin/MainActivity.kt
package com.rudderlabs.android.sample.kotlin

import android.os.Bundle
import android.widget.Button
import androidx.appcompat.app.AppCompatActivity
import com.google.android.gms.ads.identifier.AdvertisingIdClient
import com.rudderstack.android.sdk.core.RudderClient
import com.rudderstack.android.sdk.core.RudderProperty

class MainActivity : AppCompatActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        Thread {
            // This is required to register the device as a test device in the Branch dashboard
            val advertisingId = AdvertisingIdClient.getAdvertisingIdInfo(this).id
            println("AdvertisingId: $advertisingId")
        }.start()

        findViewById<Button>(R.id.identify).setOnClickListener { identify() }

        // Commerce Events
        findViewById<Button>(R.id.productAdded).setOnClickListener { productAdded() }
        findViewById<Button>(R.id.productAddedToWishList).setOnClickListener { productAddedToWishList() }
        findViewById<Button>(R.id.cartViewed).setOnClickListener { cartViewed() }
        findViewById<Button>(R.id.checkoutStarted).setOnClickListener { checkoutStarted() }
        findViewById<Button>(R.id.paymentInfoEntered).setOnClickListener { paymentInfoEntered() }
        findViewById<Button>(R.id.orderCompleted).setOnClickListener { orderCompleted() }
        findViewById<Button>(R.id.promotionViewed).setOnClickListener { promotionViewed() }
        findViewById<Button>(R.id.promotionClicked).setOnClickListener { promotionClicked() }
        findViewById<Button>(R.id.reserve).setOnClickListener { reserve() }

        // Content Events
        findViewById<Button>(R.id.productsSearched).setOnClickListener { productsSearched() }
        findViewById<Button>(R.id.productViewed).setOnClickListener { productViewed() }
        findViewById<Button>(R.id.productListViewed).setOnClickListener { productListViewed() }
        findViewById<Button>(R.id.productReviewed).setOnClickListener { productReviewed() }
        findViewById<Button>(R.id.productShared).setOnClickListener { productShared() }
        findViewById<Button>(R.id.initiateStream).setOnClickListener { initiateStream() }
        findViewById<Button>(R.id.completeStream).setOnClickListener { completeStream() }

        // Lifecycle Events
        findViewById<Button>(R.id.completeRegistration).setOnClickListener { completeRegistration() }
        findViewById<Button>(R.id.completeTutorial).setOnClickListener { completeTutorial() }
        findViewById<Button>(R.id.achieveLevel).setOnClickListener { achieveLevel() }
        findViewById<Button>(R.id.unlockAchievement).setOnClickListener { unlockAchievement() }
        findViewById<Button>(R.id.invite).setOnClickListener { invite() }
        findViewById<Button>(R.id.login).setOnClickListener { login() }
        findViewById<Button>(R.id.startTrial).setOnClickListener { startTrial() }
        findViewById<Button>(R.id.subscribe).setOnClickListener { subscribe() }

        // Custom Events
        findViewById<Button>(R.id.customTrackWithoutProperties).setOnClickListener { customTrackWithoutProperties() }
        findViewById<Button>(R.id.customTrackWithProperties).setOnClickListener { customTrackWithProperties() }

        // Reset
        findViewById<Button>(R.id.reset).setOnClickListener { reset() }
    }

    private fun identify() {
        RudderClient.getInstance()?.identify("some_user_id")
    }

    // Commerce Events

    private fun productAdded() {
        RudderClient.getInstance()?.track("Product Added", getSingleProductProperties())
    }

    private fun productAddedToWishList() {
        RudderClient.getInstance()?.track("Product Added to Wishlist", getCustomProperties())
    }

    private fun cartViewed() {
        RudderClient.getInstance()?.track("Cart Viewed", getMultipleProductProperties())
    }

    private fun checkoutStarted() {
        RudderClient.getInstance()?.track("Checkout Started", getMultipleProductProperties())
    }

    private fun paymentInfoEntered() {
        RudderClient.getInstance()?.track("Payment Info Entered", getMultipleProductProperties())
    }

    private fun orderCompleted() {
        RudderClient.getInstance()?.track("Order Completed", getMultipleProductProperties())
    }

    private fun promotionViewed() {
        RudderClient.getInstance()?.track("Promotion Viewed", getCustomProperties())
    }

    private fun promotionClicked() {
        RudderClient.getInstance()?.track("Promotion Clicked", getCustomProperties())
    }

    private fun reserve() {
        RudderClient.getInstance()?.track("Reserve", getCustomProperties())
    }

    // Content Events

    private fun productsSearched() {
        RudderClient.getInstance()?.track("Products Searched", getEComPropertiesWithoutProducts())
    }

    private fun productViewed() {
        RudderClient.getInstance()?.track("Product Viewed", getSingleProductProperties())
    }

    private fun productListViewed() {
        RudderClient.getInstance()?.track("Product List Viewed", getMultipleProductProperties())
    }

    private fun productReviewed() {
        RudderClient.getInstance()?.track("Product Reviewed", getSingleProductProperties())
    }

    private fun productShared() {
        RudderClient.getInstance()?.track("Product Shared", getSingleProductProperties())
    }

    private fun initiateStream() {
        RudderClient.getInstance()?.track("Initiate Stream", getCustomProperties())
    }

    private fun completeStream() {
        RudderClient.getInstance()?.track("Complete Stream", getCustomProperties())
    }

    // Lifecycle Events

    private fun completeRegistration() {
        RudderClient.getInstance()?.track("Complete Registration", getCustomProperties())
    }

    private fun completeTutorial() {
        RudderClient.getInstance()?.track("Complete Tutorial", getCustomProperties())
    }

    private fun achieveLevel() {
        RudderClient.getInstance()?.track("Achieve Level", getCustomProperties())
    }

    private fun unlockAchievement() {
        RudderClient.getInstance()?.track("Unlock Achievement", getCustomProperties())
    }

    private fun invite() {
        RudderClient.getInstance()?.track("Invite", getCustomProperties())
    }

    private fun login() {
        RudderClient.getInstance()?.track("Login", getCustomProperties())
    }

    private fun startTrial() {
        RudderClient.getInstance()?.track("Start Trial", getCustomProperties())
    }
    private fun subscribe() {
        RudderClient.getInstance()?.track("Subscribe", getCustomProperties())
    }

    // Custom Events

    private fun customTrackWithoutProperties() {
        RudderClient.getInstance()?.track("Custom Track Without Properties")
    }

    private fun customTrackWithProperties() {
        RudderClient.getInstance()
            ?.track("Custom Track With Properties", appendCustomProperties(RudderProperty()))
    }

    private fun reset() {
        RudderClient.getInstance()?.reset(true)
    }

    private fun getMultipleProductProperties(): RudderProperty {
        val product1 = mapOf(
            "price" to 154.79,
            "currency" to "USD",
            "brand" to "some_brand_1",
            "name" to "some_name_1",
            "category" to "Animals & Pet Supplies", // Make sure this is a valid value, refer to the Branch "ProductCategory" class.
            "sku" to "some_sku_1",
            "product_id" to "some_product_id_1",
            "quantity" to 10.0,
            "variant" to "some_variant_1",
            "rating" to 4.5
        )
        val product2 = mapOf(
            "price" to 254.79,
            "currency" to "INR",
            "brand" to "some_brand_2",
            "name" to "some_name_2",
            "category" to "Animals & Pet Supplies", // Make sure this is a valid value, refer to the Branch "ProductCategory" class.
            "sku" to "some_sku_2",
            "product_id" to "some_product_id_2",
            "quantity" to 20.0,
            "variant" to "some_variant_2",
            "rating" to 3.5
        )

        val products = listOf(product1, product2)

        val property = RudderProperty()
            .putValue("products", products)

        appendStandardProperties(property)
        return appendCustomProperties(property)
    }

    private fun getSingleProductProperties(): RudderProperty {
        val property = RudderProperty()
            .putValue("price", 154.79)
            .putValue("currency", "USD")
            .putValue("brand", "some_brand")
            .putValue("name", "some_name")
            .putValue("category", "some_category")
            .putValue("sku", "some_sku")
            .putValue("product_id", "some_product_id")
            .putValue("quantity", 10.0)
            .putValue("variant", "some_variant")
            .putValue("rating", 4.5)
        return appendCustomProperties(property)
    }

    private fun getEComPropertiesWithoutProducts(): RudderProperty {
        val property = RudderProperty()
        appendStandardProperties(property)
        return appendCustomProperties(property)
    }

    private fun appendStandardProperties(property: RudderProperty) =
        property
            .putValue("affiliation", "some_affiliation")
            .putValue("currency", "USD")
            .putValue("coupon", "some_coupon")
            .putValue("revenue", 754.79)
            .putValue("shipping", 100.0)
            .putValue("tax", 5.0)
            .putValue("order_id", "some_order_id")
            .putValue("query", "some_query")


    private fun getCustomProperties() = appendCustomProperties(RudderProperty())
    private fun appendCustomProperties(property: RudderProperty) =
        property
            .putValue("key-1", "value-1")
            .putValue("key-2", 123)
}
